### PR TITLE
TF-1008: Error when deleting mailbox

### PR DESCRIPTION
**Pull request by ManhNTX**
Need PR Merge : https://github.com/linagora/jmap-dart-client/pull/48

https://user-images.githubusercontent.com/107173849/199377294-5d227a75-7b2a-4781-8d09-2c4047e03c37.mov

Screen Shot 2022-11-04 at 14 16 33

### CODE REVIEW

File: lib/features/mailbox/data/model/mailbox_not_destroyed_type.dart [hunk 1]
```diff
@@ -0,0 +1,3 @@
+enum MailBoxNotDestroyedType {
+  notFound,
+}
```
Comment (hoangdat) [inline lib/features/mailbox/data/model/mailbox_not_destroyed_type.dart:3]: `jmap-dart-client` have not defined the type for this error before?

Comment (ManhNTX) [reply lib/features/mailbox/data/model/mailbox_not_destroyed_type.dart:3]: @hoangdat No

File: lib/features/mailbox/data/network/mailbox_api.dart [hunk 2]
```diff
@@ -193,13 +195,19 @@ class MailboxAPI {
...
         .every((element) => element == true);
     }
 
     return finalResult;
   }
 
+  bool _validateMailBoxDestroyedSuccess(SetMailboxResponse? response) {
+    return response?.destroyed?.isNotEmpty == true ||
+      response?.notDestroyed?.values
+        .where((e) => e.type == ErrorType(MailBoxNotDestroyedType.notFound.name)).isNotEmpty == true;
```
Comment (dab246) [inline lib/features/mailbox/data/network/mailbox_api.dart:208]: IMO, Such checks should not be made. We will miss the delete failure cases.

File: lib/features/mailbox/data/network/mailbox_api.dart [hunk 3]
```diff
@@ -193,11 +197,21 @@ class MailboxAPI {
 
       finalResult = currentSetMailboxInvocations
         .map((currentInvocation) => response.parse(currentInvocation.methodCallId, SetMailboxResponse.deserialize))
-        .map((response) => response?.destroyed?.isNotEmpty ?? false)
-        .every((element) => element == true);
+        .map((response) {
+          if(response?.notDestroyed != null) {
+            deleteMailboxErrors.addAll(response!.notDestroyed!);
```
Comment (hoangdat) [inline lib/features/mailbox/data/network/mailbox_api.dart:202]: `response` here is non-null. please carefully

Comment (ManhNTX) [reply lib/features/mailbox/data/network/mailbox_api.dart:202]: check at line 201

Comment (hoangdat) [reply lib/features/mailbox/data/network/mailbox_api.dart:202]: it is not a check for `response`. Still risk when using `response!`

File: lib/features/mailbox/data/mixin/error_handle_mixin.dart [hunk 4]
```diff
@@ -0,0 +1,21 @@
+import 'package:jmap_dart_client/jmap/core/error/set_error.dart';
+import 'package:jmap_dart_client/jmap/core/id.dart';
+
+mixin ErrorHandleMixin {
+  void remoteHandleError({
+    required Map<Id,SetError> errors,
+    Function(Map<Id,SetError>)? handleNotFoundError,
+    Function(Map<Id,SetError>)? handleUnKnowErrorNotFound,
```
Comment (hoangdat) [inline lib/features/mailbox/data/mixin/error_handle_mixin.dart:8]: how about if we want to add more handler for other kind of error?

File: lib/features/mailbox/data/mixin/error_handle_mixin.dart [hunk 5]
```diff
@@ -0,0 +1,21 @@
+import 'package:jmap_dart_client/jmap/core/error/set_error.dart';
+import 'package:jmap_dart_client/jmap/core/id.dart';
+
+mixin ErrorHandleMixin {
+  void remoteHandleError({
```
Comment (hoangdat) [inline lib/features/mailbox/data/mixin/error_handle_mixin.dart:5]: handleError

File: lib/features/mailbox/data/mixin/error_handle_mixin.dart (hunk 4 above)
Comment (hoangdat) [reply lib/features/mailbox/data/mixin/error_handle_mixin.dart:8]: it is not follow open-closed principle

File: model/lib/error_type_handler/error_type_handler.dart [hunk 6]
```diff
@@ -0,0 +1,18 @@
+import 'package:jmap_dart_client/jmap/core/error/error_type.dart';
+import 'package:jmap_dart_client/jmap/core/error/set_error.dart';
+import 'package:jmap_dart_client/jmap/core/id.dart';
+typedef Error = Map<Id, SetError>;
```
Comment (dab246) [inline model/lib/error_type_handler/error_type_handler.dart:4]: IMO, such generic names should not be given. Could be `JmapError ...v.v.`

**Comment by hoangdat** [comment]
please refer: https://github.com/linagora/tmail-flutter/pull/1174.

**Pull request closed without merging**